app/src/main/java/pm/gnosis/heimdall/ui/safe/helpers/RecoverSafeOwnersHelper.kt

```diff
@@ -163,7 +165,43 @@ class DefaultRecoverSafeOwnersHelper @Inject constructor(
             payload
         }
 
-        return when (payloads.size) {
+        return buildTransaction(safeInfo, payloads)
+    }
+
+    override fun buildRecoverTransaction(
+        safeInfo: SafeInfo,
+        addressesToKeep: List<Solidity.Address>,
+        addressesToAdd: List<Solidity.Address>
+    ): SafeTransaction =
+        buildTransaction(safeInfo, buildPayloads(safeInfo, addressesToKeep, addressesToAdd))
+
+    private fun buildPayloads(
+        safeInfo: SafeInfo,
+        addressesToKeep: List<Solidity.Address>,
+        addressesToAdd: List<Solidity.Address>
+    ): List<String> {
+        val payloads = mutableListOf<String>()
+        val remainingNewAddresses = addressesToAdd.toMutableList()
+        for (i in safeInfo.owners.size - 1 downTo 0) {
+            val safeOwner = safeInfo.owners[i]
+            if (remainingNewAddresses.isEmpty()) break
+            if (!addressesToKeep.contains(safeOwner)) {
+                val newOwner = remainingNewAddresses.removeAt(remainingNewAddresses.size - 1)
+                if (safeInfo.owners.contains(newOwner)) continue
```

Comment:
Is this correct?

we will also try the next safe owner, since the continue advances the for loop, right?

Reply:
If `newOwner` is already present in the `safeInfo` then we don't need to add it and we continue the loope.